// @ant-basement/miniprogram-sdk/dist/esm/basement.js
import * as tslib_1 from "tslib";
import { Basement as Base, WHITELIST_EXTENSIONS, assert, OSSUploadHeaderList, BasementClientError, ErrorType, ErrorName, ErrorCode, OSSUploadResponseFormat, } from '@ant-basement/core';
import { DbService, FileService, AuthService, FunctionService, AppService, } from '@ant-basement/services';
import { MiniProgramHTTPTransport } from './transport';
import mime from 'mime/lite';
var MiniProgramFileService = (function (_super) {
    tslib_1.__extends(MiniProgramFileService, _super);
    function MiniProgramFileService() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MiniProgramFileService.prototype.uploadFile = function (options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var relativePath, extension, meta, headers, uploadRes, uploadOptions;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        assert(options.filePath && typeof options.filePath === 'string', 'missing options.filePath');
                        relativePath = options.filePath.replace(/(.*):\/\//, '');
                        extension = relativePath.split('.').pop();
                        assert(WHITELIST_EXTENSIONS.includes(extension.toLowerCase()), "\u76EE\u524D\u4E0D\u652F\u6301 " + extension + " \u7C7B\u578B\u6587\u4EF6");
                        meta = Object.keys(options.meta || {}).reduce(function (accu, key) {
                            accu["x-oss-meta-" + key] = options.meta[key];
                            return accu;
                        }, {});
                        headers = options.headers ? OSSUploadHeaderList.reduce(function (accu, key) {
                            var fieldName = key.replace(/\-[A-Z]/g, function (match) { return match[1]; }).replace(/^[A-Z]/, function (match) { return match.toLowerCase(); });
                            if (options.headers.hasOwnProperty(fieldName))
                                accu[key] = options.headers[fieldName];
                            return accu;
                        }, {}) : {};
                        return [4, this.getOSSUploadOptionsFromPath(relativePath, options.path, options.fileSize)];
                    case 1:
                        uploadRes = _a.sent();
                        if (uploadRes.error) {
                            throw new BasementClientError(ErrorName.INTERFACE_ERROR, ErrorCode.INTERFACE_RESPONSE_FAILED, ErrorType.COMMON_ERROR, uploadRes.error.message);
                        }
                        uploadOptions = OSSUploadResponseFormat(uploadRes.result);
                        return [4, this.uploadFileToOSS(options, uploadOptions, headers, meta)];
                    case 2:
                        _a.sent();
                        return [4, this.reportOSSUpload(uploadOptions.id, mime.getType(options.extension))];
                    case 3:
                        _a.sent();
                        return [2, {
                                fileUrl: "https://" + uploadOptions.host + "/" + uploadOptions.key,
                                filePath: uploadOptions.key,
                            }];
                }
            });
        });
    };
    MiniProgramFileService.prototype.uploadFileToOSS = function (fileUploadOptions, ossUploadOptions, headers, meta) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var options, uploadHeader, contentType;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        options = ['key', 'policy', 'Signature', 'OSSAccessKeyId'].reduce(function (accu, key) {
                            accu[key] = ossUploadOptions[key];
                            return accu;
                        }, ossUploadOptions);
                        uploadHeader = {};
                        if (fileUploadOptions.extension) {
                            contentType = mime.getType(fileUploadOptions.extension);
                            if (!contentType) {
                                throw new BasementClientError(ErrorName.VALIDATION_ERROR, ErrorCode.VALIDATION_FAILED, ErrorType.COMMON_ERROR, '文件扩展错误，无法解析正确的 MIME');
                            }
                            uploadHeader['Content-Type'] = contentType;
                        }
                        headers['Cache-Control'] = 'max-age=2592000';
                        return [4, this.transport.upload("https://" + ossUploadOptions.host, Object.assign({ success_action_status: 200 }, headers, meta, options), 'file', fileUploadOptions.filePath, uploadHeader)];
                    case 1:
                        _a.sent();
                        return [2];
                }
            });
        });
    };
    return MiniProgramFileService;
}(FileService));
var Basement = (function (_super) {
    tslib_1.__extends(Basement, _super);
    function Basement(appGlobal, options) {
        var _this = _super.call(this, tslib_1.__assign({}, options, { httpClient: appGlobal, httpTransport: MiniProgramHTTPTransport })) || this;
        assert(options.clientSecret, 'client secret is required');
        if (Object.keys(options).length === 0) {
            console.error('云服务配置不存在');
            return _this;
        }
        if (!options.appId) {
            appGlobal.showToast({ content: '请检查是否选择应用' });
            return _this;
        }
        if (!options.spaceId) {
            appGlobal.showToast({ content: '请检查是否选择云服务' });
            return _this;
        }
        _this.db = new DbService(_this.transport);
        _this.user = new AuthService(_this.transport);
        _this.file = new MiniProgramFileService(_this.transport);
        _this.function = new FunctionService(_this.transport);
        _this.appService = new AppService(_this.transport);
        if (appGlobal.remoteLog) {
            appGlobal.remoteLog({
                type: 'behavior',
                seedId: 'basementVersion',
                param1: _this.version,
                bizType: 'basement',
                logLevel: 1,
                actionId: 'event',
            });
        }
        return _this;
    }
    Object.defineProperty(Basement.prototype, "version", {
        get: function () {
            return VERSION;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Basement.prototype, "ua", {
        get: function () {
            return "pkg_name:@ant-basement/miniprogram-sdk;ver:" + this.version + ";";
        },
        enumerable: true,
        configurable: true
    });
    Basement.prototype.createTransport = function (options) {
        _super.prototype.createTransport.call(this, options);
        this.transport.setAppSecret(options.clientSecret).setUA(this.ua);
    };
    return Basement;
}(Base));
export { Basement };
