// @ant-basement/miniprogram-sdk/dist/umd/basement.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "tslib", "@ant-basement/core", "@ant-basement/services", "./transport", "mime/lite"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const tslib_1 = require("tslib");
    const core_1 = require("@ant-basement/core");
    const services_1 = require("@ant-basement/services");
    const transport_1 = require("./transport");
    const lite_1 = tslib_1.__importDefault(require("mime/lite"));
    class MiniProgramFileService extends services_1.FileService {
        async uploadFile(options) {
            core_1.assert(options.filePath && typeof options.filePath === 'string', 'missing options.filePath');
            const relativePath = options.filePath.replace(/(.*):\/\//, '');
            const extension = relativePath.split('.').pop();
            core_1.assert(core_1.WHITELIST_EXTENSIONS.includes(extension.toLowerCase()), `目前不支持 ${extension} 类型文件`);
            const meta = Object.keys(options.meta || {}).reduce((accu, key) => {
                accu[`x-oss-meta-${key}`] = options.meta[key];
                return accu;
            }, {});
            const headers = options.headers ? core_1.OSSUploadHeaderList.reduce((accu, key) => {
                const fieldName = key.replace(/\-[A-Z]/g, match => match[1]).replace(/^[A-Z]/, match => match.toLowerCase());
                if (options.headers.hasOwnProperty(fieldName))
                    accu[key] = options.headers[fieldName];
                return accu;
            }, {}) : {};
            const uploadRes = await this.getOSSUploadOptionsFromPath(relativePath, options.path, options.fileSize);
            if (uploadRes.error) {
                throw new core_1.BasementClientError(core_1.ErrorName.INTERFACE_ERROR, core_1.ErrorCode.INTERFACE_RESPONSE_FAILED, core_1.ErrorType.COMMON_ERROR, uploadRes.error.message);
            }
            const uploadOptions = core_1.OSSUploadResponseFormat(uploadRes.result);
            await this.uploadFileToOSS(options, uploadOptions, headers, meta);
            await this.reportOSSUpload(uploadOptions.id, lite_1.default.getType(options.extension));
            return {
                fileUrl: `https://${uploadOptions.host}/${uploadOptions.key}`,
                filePath: uploadOptions.key,
            };
        }
        async uploadFileToOSS(fileUploadOptions, ossUploadOptions, headers, meta) {
            const options = ['key', 'policy', 'Signature', 'OSSAccessKeyId'].reduce((accu, key) => {
                accu[key] = ossUploadOptions[key];
                return accu;
            }, ossUploadOptions);
            const uploadHeader = {};
            if (fileUploadOptions.extension) {
                const contentType = lite_1.default.getType(fileUploadOptions.extension);
                if (!contentType) {
                    throw new core_1.BasementClientError(core_1.ErrorName.VALIDATION_ERROR, core_1.ErrorCode.VALIDATION_FAILED, core_1.ErrorType.COMMON_ERROR, '文件扩展错误，无法解析正确的 MIME');
                }
                uploadHeader['Content-Type'] = contentType;
            }
            headers['Cache-Control'] = 'max-age=2592000';
            await this.transport.upload(`https://${ossUploadOptions.host}`, Object.assign({ success_action_status: 200 }, headers, meta, options), 'file', fileUploadOptions.filePath, uploadHeader);
        }
    }
    class Basement extends core_1.Basement {
        constructor(appGlobal, options) {
            super({ ...options, httpClient: appGlobal, httpTransport: transport_1.MiniProgramHTTPTransport });
            core_1.assert(options.clientSecret, 'client secret is required');
            if (Object.keys(options).length === 0) {
                console.error('云服务配置不存在');
                return;
            }
            if (!options.appId) {
                appGlobal.showToast({ content: '请检查是否选择应用' });
                return;
            }
            if (!options.spaceId) {
                appGlobal.showToast({ content: '请检查是否选择云服务' });
                return;
            }
            this.db = new services_1.DbService(this.transport);
            this.user = new services_1.AuthService(this.transport);
            this.file = new MiniProgramFileService(this.transport);
            this.function = new services_1.FunctionService(this.transport);
            this.appService = new services_1.AppService(this.transport);
            if (appGlobal.remoteLog) {
                appGlobal.remoteLog({
                    type: 'behavior',
                    seedId: 'basementVersion',
                    param1: this.version,
                    bizType: 'basement',
                    logLevel: 1,
                    actionId: 'event',
                });
            }
        }
        get version() {
            return VERSION;
        }
        get ua() {
            return `pkg_name:@ant-basement/miniprogram-sdk;ver:${this.version};`;
        }
        createTransport(options) {
            super.createTransport(options);
            this.transport.setAppSecret(options.clientSecret).setUA(this.ua);
        }
    }
    exports.Basement = Basement;
});
