// @ant-basement/miniprogram-sdk/dist/umd/transport.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "@ant-basement/core", "./codec"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const core_1 = require("@ant-basement/core");
    const codec_1 = require("./codec");
    class MiniProgramHTTPTransport extends core_1.HTTPTransport {
        constructor(endpoint, library) {
            super(endpoint, library);
            this.scope = 'auth_user';
            core_1.assert(library.uploadFile, 'missing miniprogram execution environment');
            core_1.assert(library.getAuthCode, 'missing miniprogram execution environment');
            core_1.assert(library.canIUse, 'missing miniprogram execution environment');
            this.uploadFile = this.wrap(library.uploadFile);
            this.getAuthCode = this.wrap(library.getAuthCode);
            const httpRequest = library.canIUse('request') ? library.request : library.httpRequest;
            core_1.assert(httpRequest, 'missing miniprogram execution environment');
            this.httpRequest = this.wrap(httpRequest);
        }
        getEncoder() {
            return new codec_1.MiniProgramHTTPRequestEncoder(this.endpoint, this.spaceId);
        }
        async request(encoder, retried = false) {
            const cloned = encoder.clone();
            const token = await this.getAccessToken();
            encoder.setBodyField({
                token,
            });
            encoder.sign(this.appSecret);
            encoder.setBaseHeaders({
                'Content-Type': 'application/json',
                'x-basement-token': token,
            });
            if (this.ua) {
                encoder.setBaseHeaders({
                    'x-serverless-ua': this.ua,
                });
            }
            try {
                const encoded = encoder.encodeAsHTTPRequestObject({
                    timeout: this.timeout,
                    dataType: 'json',
                });
                this.logger.info('request data');
                this.logger.info(JSON.stringify(encoded, null, 2));
                const decoded = await this.httpRequest(encoded);
                return decoded;
            }
            catch (e) {
                const isUnAuthorized = e.error.code === 'GATEWAY_INVALID_TOKEN'
                    || e.error.code === 'InvalidParameter.InvalidToken' || e.status === core_1.HTTP_UNAUTHORIZED;
                if (isUnAuthorized) {
                    if (retried) {
                        this.logger.error('auth error', e);
                        throw new core_1.BasementClientError(core_1.ErrorName.UNAUTHORIZED_ERROR, core_1.ErrorCode.AUTHENTICATION_FAILED, core_1.ErrorType.COMMON_ERROR, 'authentication failed');
                    }
                    await this.getAccessToken(true);
                    return await this.request(cloned, true);
                }
                if (e.error) {
                    this.logger.error('IDE error', e.message);
                    this.logger.error(e);
                    throw e.error;
                }
                this.logger.error('Unknown error', e.message);
                this.logger.error(e);
                throw e;
            }
        }
        async authorize() {
            this.pendingRequest = this
                .getAuthCode({ scopes: this.scope })
                .then((res) => {
                this.logger.info('Request authcode ' + (res.body.authCode ? 'success' : 'failed'));
                return res.body.authCode;
            })
                .then((authCode) => {
                const encoder = this.getEncoder();
                encoder.setBodyField({
                    method: 'serverless.auth.user.authorize',
                    params: {
                        authProvider: 'alipay_openapi',
                        clientIdentifier: this.appId,
                        authCode,
                    },
                });
                encoder.sign(this.appSecret);
                encoder.setBaseHeaders({ 'Content-Type': 'application/json' });
                if (this.ua) {
                    encoder.setBaseHeaders({
                        'x-serverless-ua': this.ua,
                    });
                }
                const encoded = encoder.encodeAsHTTPRequestObject({
                    timeout: this.timeout,
                    dataType: 'json',
                });
                return this.httpRequest(encoded);
            })
                .then((res) => {
                this.logger.info('Request accessToken ' + (res.body.success ? 'success' : 'failed'));
                if (res.body && res.body.result) {
                    this.accessToken = res.body.result.accessToken;
                }
                this.pendingRequest = null;
                return Promise.resolve(this.accessToken);
            });
            return this.pendingRequest;
        }
        async getAccessToken(refresh = false) {
            if (this.pendingRequest) {
                this.logger.info('getAccessToken: reuse');
                return this.pendingRequest;
            }
            if (refresh || !this.accessToken) {
                this.logger.info('getAccessToken: start');
                return this.authorize();
            }
            return this.accessToken;
        }
        async upload(host, formData, fileName, filePath, header) {
            this.logger.info('upload with params');
            this.logger.info(JSON.stringify(formData, null, 2));
            return await this.uploadFile({
                url: host,
                formData,
                fileName,
                filePath,
                fileType: 'image',
                header: {
                    ...header,
                    'X-OSS-server-side-encrpytion': 'AES256',
                },
            });
        }
        wrap(myMethod) {
            return args => {
                return new Promise((resolve, reject) => {
                    myMethod(Object.assign(args, {
                        complete: (res = {}) => {
                            this.logger.info('completed request');
                            this.logger.info(JSON.stringify(res, null, 2));
                            const decoder = new codec_1.MiniProgramHTTPResponseDecoder();
                            const response = decoder.decode(res);
                            if (response.error) {
                                return reject(response);
                            }
                            return resolve(response);
                        },
                    }));
                });
            };
        }
    }
    exports.MiniProgramHTTPTransport = MiniProgramHTTPTransport;
});
